### ISSUE TITLE
Update to enable Pillar 2 sample IDs to be shared via CLIMB

### ISSUE THREAD

**Issue by tomrconnor** [COG-UK/dipi-group#78]
**Issue:**

At present Pillar 2 samples are sequenced in a number of locations within the UK including Wellcome Sanger, Public Health Wales and in Northern Ireland (by HSCI/Belfast Trust/Queens).

At present, Pillar 2 sequences are uploaded to CLIMB, but these are provided without sample metadata which has to be matched and uploaded by the PHA that covers the region where the case is resident. This process is currently effectively manual, with an exchange of CSVs via an sFTP, which can result in delays ranging from 2 days to 2 weeks in some circumstances. Time delays vary across the UK, with the system currently providing an inequitable service.

At present this system also makes it difficult to track and identify 'true' orphan samples, an area that is becoming an increasing concern for some PHAs.

The core use case for this requirement (one site doing sequencing and a separate agency providing the metadata) already exists within COG-UK/CLIMB COVID for Pillar 1, and so this provides a potential template/basis for the implementation of a solution for Pillar 2 sequencing.

**Proposed Solution**:

Being able to use a similar process to the pillar 1 system that is currently in place for England, where Pillar 2 sequencing sites upload a sample ID which enables a PHA to match the sample to patient records and generate the required metadata for upload would reduce the time to match metadata for Wales, NI and Scotland significantly.

Stage 1: Pillar 2 sequencing site uploads sequence data that they are unable to match to their own records/data to CLIMB COVID and uploads associated sequencing information into Majora

Stage 2: each PHA queries the list of 'unclaimed' sample IDs (on a regular basis - in PHW this would be daily), looks for matches against their local records, and 'claims' any samples that are from patients resident in their country. The PHA then generates the sample metadata and uploads this to CLIMB COVID
    - **_For IG/GDPR reasons it may be that once a PHA claims a sample ID, there is the option to remove/blank out that sample ID from the system_**

Stage 3: Elan/analysis pipelines run as normal, and PHAs are able to pull back the analysis results in a timely way. 
   - **_Versions of the PHW variant pipeline may be required/desired for NI/Scotland to automate this process_**

Stage 4: Unclaimed pillar 2 samples are tracked centrally, with all PHAs having visibility on them, and where a sample ID has been unclaimed for longer than an agreed period of time it is considered an 'orphan'

**Timeline:**

Ideally this system would be in place/in use asap. Happy to discuss the specifics, and appreciate that this would require some work from Sanger and others to implement. Propose running this in parallel to the current sFTP solution for validation, before switching off the sFTP route. 

**Approvals:**

The proposed changes were conceived and agreed at the Data Integration working group of the UK COVID-19 Genomics Strategic Advisory Board, and have been presented, discussed and agreed at the COG-UK Health Informatics group. Representatives from all 4 PHAs have agreed to, and support, these changes/additional features. 

**Summary:**

There are some significant delays in getting Pillar 2 sequence data back to PHAs where it can be actioned. The current semi-manual process that makes use of sFTP is slow and introduces unnecessary delays. The envisaged functionality (which builds on what is there already - and may already fully exist within CLIMB COVID) will allow a significant improvement of turnaround times for sample to action.

**Comment by SamStudio8** [comment]
We can offer most of this request already; as you know we have a method in place for sharing the precious `sender_sample_id` for English Pillar 1 samples with PHE and it will be trivial to extend this to cover the other nations. Each PHA will need to familiarise itself with Ocarina, the [documentation for using dataviews](https://docs.covid19.climb.ac.uk/accessing-dataviews) and [setting up an OAuth application on Majora](https://docs.covid19.climb.ac.uk/oauth-app). FS and WSI users have already done this so will be able to offer some assistance.

The tricky part comes with how to get the identifiers into CLIMB-COVID. This is quite straightforward for Pillar 1, as the mandatory metadata is (mostly) available and the submitting sites can create the biosample artifacts in Majora (with the `sender_sample_id`) themselves. Pillar 2 has the more unusual problem of having the sequence, but almost no metadata other than the `sender_sample_id`. Which leaves them unable to create a biosample artifact (even a bare minimum one -- as they likely don't even know which `adm1` is correct) to attach the linkage to. 

Luckily, we do already have a system in place for creating "blank" biosample artifacts. This was (is still?) being used by the WSI to ensure sequences could be uploaded by downstream sequencing sites when pushing overflow Pillar 2 samples into the COG-UK sequencing network, despite biosample metadata not yet being available.

As the WSI is already able to use this procedure, It seems to me that overriding this "blank" procedure would be the easiest way to achieve what you've outlined here. Pillar 2 sequencing sites can force push a blank biosample (and we can somehow allow for a `sender_sample_id` to be attached here), and those biosamples will still be eligible to appear in the dataviews that return the `central_sample_id`-`sender_sample_id` linkage to PHAs.

All PHAs will then be able to use Ocarina to securely fetch the linkage as often as they like, directly from Majora -- the same way that PHE does so now. Claimed biosamples can be tagged with Ocarina (the same way they are now) which will updated the key-value flags and help keep track of which PHA has claimed a biosample artifact. PHAs are also welcome to use Ocarina's partial update feature to "stomp" over the `sender_sample_id` if they want to linkage to be removed for IG reasons. Everything described in this paragraph is already operational.

This will allow for orphans to be tracked centrally, as anyone in the consortium can query the daily metadata table to check the flags from each PHA. An analyst could use the `sequencing_submission_date`, to determine whether the biosample is old enough to be an orphan without any additional work from us.

Action:
* SN to update the process for force pushing blank biosamples to allow for `sender_sample_id` to be added (perhaps a discussion for `#metadata-apis` to ensure any changes are suitable for users)
    * Requiring a change in the Majora API for blank biosamples,
    * some documentation (this interface is not documented at all currently), and
    * a little bit of work on WSI side to feed the IDs to the new API interface
* Dataviews will need to be added for each PHA who wants to pull linkage (to show their own adm1 samples, and adm1 null samples)
* PHAs who want to pull linkage will need to register an OAuth application and get Ocarina up and running on their end
* We'll need to draft some guidance on how all this will work (with example Ocarina commands etc.)

Timeline:
I anticipate the bulk of this work could be implemented by the end of this week, with PHAs operating on the new method by the end of next week. Update: I can't speak for WSI of course.

**Comment by SamStudio8** [comment]
Although some more careful scoping and testing will be needed to avoid [what happened last time we updated the blank biosample artifact interface](https://github.com/COG-UK/dipi-group/blob/official/postmortems/20210305_blankbiosample.md)...

**Comment by cs46** [comment]
@SamStudio8 We do have code setup to create a blank sample but at the moment this is not being used at all as it was only for when we sent sample for external sequencing. We needed to create a blank biosample record for them as they didnt have permissions to do so and then they could add their metadata.  I believe that these sites now have a sFTP connection to PHE so this hasn't been utilised in a few months and due to the decrease in positives we have not sent any off site in a while.

Our existing code base expects us to have metadata which is obviously not compatible with the above proposal.  We could modify the existing metadata upload to include this extra field (minimal work) but would only be compatible with English samples as we do not have data for DAs. We would then need a separate method to create a blank biosample record with just the sender sample id filled for any other new sequence uploaded where we dont hold metadata.  There are hacky ways of doing the latter to ensure we could put something in place quickly and other options which will take us longer to put in place but it would be more robust.  With this in mind we should definitely run the current sFTP processes in parallel for the foreseeable future.

We also need to ensure that the /api/v2/artifact/biosample/query/validity/ API call would still work as intended if we make this change as it is what we use to determine if we need to do something with the sample metadata.  Would the "has_metadata" flag still be shown as "FALSE" if we only uploaded the sender sample id?

Additionally, we should have a method to state that we have created the record to provide linkage data only so that there is no confusion as to why a record has been created.

**Comment by cs46** [comment]
@tomrconnor stupid question but are you expecting the lighthouse lab bar code to be entered into the sender_sample_id field rather than a Sanger generated code (the latter of which doesn't exist)?  I only ask as the definitions in the consortium agreement would lead me to believe the sender sample ID was originally for a sequencing site to upload their internally generated code rather than what we are proposing to use it for.  Fine with the change but i want to be clear that i have understood the ask.

There is also the root_sample_id field which is used interchangeable with the bar code (either rightly or wrongly) which is confusing matters.

**Comment by SamStudio8** [comment]
@cs46 Thanks, sounds like this is where the core of the work will be. I think the meat of this proposal is to ensure the process works with equity for all four nations so we'll need a process that works for English and non-English samples. My broad stroke recommendation would be that WSI should not need to provide any metadata at all -- even for English samples, which should come from PHE after running this new process. I'd definitely vote towards building something robust to cover all the DAs here and happy to help facilitate as needed, we can run the sFTP method in the meantime as usual.

The validity call currently returns `False` for `has_metadata` when the sample is "blank", and should continue to do so, even for these "blank plus sender_sample_id" artifacts. It should be clear from the current level of auditing that a blank sample was created with a `sender_sample_id` that the sample was created for this purpose, so I think that is covered too.

re: your question to @tomrconnor, the `sender_sample_id` needs to be populated with something the DAs can actually link locally, so WSI-specific identifiers are not going to work. The CA explicitly mentions these internal identifiers can be shared to CLIMB-COVID in this way and should cover the lighthouse identifiers (the original definition asked that this be the ID that was shared with PHE via SGSS, but things have been blurred a little since). On `root_sample_id`, those are specifically provided by DAs only -- but may well be the same as the `sender_sample_id` in some cases. Not sure if any of that helps but I'm sure TRC can clarify.

**Comment by tomrconnor** [comment]
@cs46 yep, we would want the lighthouse lab barcode in the sender_sample_id field. We can then match the lighthouse lab ID to data on our LIMS. 

The sender_sample_id field has been used (i think exclusively to date) to provide the NHS lab ID (or equivalent) so that PHE can match the case to the patient data that they hold; and that is basically what we want to do here, but the lighthouse lab IDs are UK-wide/universal - so they can be used by everyone.

Does that make sense?

**Comment by SamStudio8** [comment]
After a discussion with FS and RM about PHE linkage generally, it sounds like our proposal to modify the blank biosample to allow for sender_sample_id  would neatly solve an outstanding issue with "Pillar 103" (Pillar 2 commercial samples). I'll begin drafting a solution to add this to Majora. The current PHE pipeline is already primed to take the secure dataview and create/update Pillar 2 samples so can easily take over for Pillar 103.

**Comment by SamStudio8** [comment]
For the most minimal amount of work, we could set an optional `sender_sample_id` in here https://github.com/SamStudio8/majora/blob/stable/majora2/api_views.py#L839; and make `sender_sample_id` an optional field when pushing blanks to the `api.artifact.biosample.addempty` API endpoint. This would allow the biosample to be observed through the secure dataviews, but not through anything else (as desired).

The only catch (which actually seems like an upside, given Tom's suggestion about IG), is that `--partial` updates cannot be invoked on these empty biosamples (https://github.com/SamStudio8/majora/blob/stable/majora2/api_views.py#L925) which means they'd have to be stomped into existence -- which would lose the `sender_sample_id` unless it was explicitly repushed in the call to create the biosample. This is actually fine for our use-case, as we technically don't need to keep the `sender_sample_id` in Majora once the linkage has been established, but is worth nothing nonetheless.

I don't like what a hack this endpoint turned out to be, but it works, and we don't have the time to do something more elegant. I guess medium-to-long term we'd incorporate the possibility that an artifact is incomplete into the core `MajoraArtifact` model instead.

**Comment by cs46** [comment]
@SamStudio8 i also need to understand how to update the is_surveillance flag.  We have started to receive the sample of interest data yesterday which we are monitoring to understand its contents.  We would be in a position to upload this at the same time as the lab bar code for all English samples.  If not we would have to write special logic which would check to see if it had been "stomped" and then update it after the fact using partial.  It would be good if we could upload both at the same time to avoid this complexity.

For all non english samples we would perhaps need to set them to a default "N" and leave them for the DAs to update with their information (or just leave it blank).

**Comment by SamStudio8** [comment]
@cs46 You wouldn't upload it at all, would be my recommendation. Do the PHAs not have this information to flag them afterward?

**Comment by cs46** [comment]
@SamStudio8 not to my knowledge but I'm happy to be proven wrong.

**Comment by SamStudio8** [comment]
@cs46 Damn. I'll ask around as it would be good to know whether it is just hard to find out, or impossible to find out on their end. Seems to me that all metadata should just come from the canonical PHA sources. My plan would be to release WSI from having to worry about metadata at all. Supporting a use-case wherein the supposedly blank biosample artifacts actually have a bunch of things like is_surveillance in there is going to be much more tricky and I see problems in the long run for that sort of hackery.

**Comment by cs46** [comment]
@SamStudio8 i will ask in the PHE meeting later to see if they do/can get access to the data. We are probably the one of the first to be given access so no one else will have this information.

There is also an argument for capacity from PHEs side and our service is here to help out as required.  I agree it would be best if the PHAs did all of the linking but we must also consider resilience, should we still have an upload process in case the PHE process fails?  This is going off topic a little bit but we are in this grey area now and we will have to come to a decision on who owns what and where Sanger fits into the picture.  I would advocate that we act more as a backup for the core metadata upload and as a primary source when we have been given access to certain data first.  I dont believe we can have a blanket PHA does everything just yet.

**Comment by SamStudio8** [comment]
@cs46 Time for a real meeting I suspect ;)

**Comment by SamStudio8** [comment]
https://github.com/SamStudio8/majora/commit/373aed958e32416cd23195e22b657978eee0ce99 adds `sender_sample_id` as an optional key in the `addempty` API endpoint (as previously described). Ocarina 0.40.0 adds `--sender-sample-id` as an option to `empty biosample` https://github.com/SamStudio8/ocarina/commit/ffe4b9b024ae7b4d1639cca0a2c6d41b11c3b8d5.

This has been deployed and is working in Majora Magenta. I am liaising with @cs46 to test this for their process before deploying to Majora actual.

**Comment by SamStudio8** [comment]
https://github.com/SamStudio8/majora/commit/feef645648aebfdbc624867b127e856a0e7efb38 (and others) add a significant test suite to cover new functionality in the hope that we won't blow it up like last time. Ocarina has been modified further (https://github.com/SamStudio8/ocarina/commit/ff03bbf3615c37ef393ff7ea72acecfb23b2c868) and 0.40.2 meets the implemented spec in Majora. The updates are live in Majora Magenta.

**Comment by SamStudio8** [comment]
After spending more time than I care to admit hacking the Majora Dataviews to support fetching samples with their `adm1` set to `None` (* https://github.com/SamStudio8/majora/commit/0546145d265448896a5ee857ab9aeb64e5ccabda), I realised that the existing dataviews will not work for blank samples, as the filter definitions also require the uploading user to have signed off on sharing the `sender_sample_id`. However, the uploading user for blank samples is -- you guessed it -- blank.

This is the wrong way to go about this anyway, so instead I've created a new dataview that can be shared across all the DAs, which give all members access to any `central_sample_id` and `sender_sample_id` pair where `adm1` is `None`. That is, a dataview where the linkage is basically up for grabs. So long as we tightly control users who are allowed to create blank samples (which can be done as it requires a particular OAuth scope), we don't need to think about adding this into the Pillar 1 permission framework.

(*) This was complicated as `filter(adm1__in=[None, "UK-ENG"])` does not work. The None cannot be transformed into a comparable value when it comes to building the queries SQL, necessitating an additional `filter(adm1__isnull=True)`. Except these now clash with one another and require a Django `Q` object to permit consideration of both queries as an `OR`. Not only was the existing filtering method (written by yours truly) unable to deal with `None` values, it didn't use anything fancy like `Q` objects either, opting to return a mere dict that gets dumped into `filter()` later. Luckily building them is pretty straightforward and `MajoraDataview.get_filters` now returns a `Q` object instead of a `dict` (and adds an extra `isnull` `Q` if needed). Although as it happens this was all pointless because coming up with a distinct dataview for orphans was a much better idea in the first place.

(This is all still on Majora Magenta)

**Comment by SamStudio8** [comment]
Tested this with @cs46 on Magenta :+1: CS to make a new script to handle collection_pillar and other gubbins for English samples.

**Comment by SamStudio8** [comment]
* [x] Chase approval
* [x] Notify DIPI (https://github.com/COG-UK/dipi-group/commit/df30a648d9f9573a52f2b3a45d5eb5f35a1cce1c)
* [x] Majora database migration
* [x] Update Ocarina if needed (https://github.com/SamStudio8/ocarina/commit/ffe4b9b024ae7b4d1639cca0a2c6d41b11c3b8d5)
* [x] API docs (https://github.com/SamStudio8/majora-docs/commit/eea83df90a18c9904a511597efbd0ef5dea964c3)
* [x] Push to Majora prod
* [x] Majora production database migration

Changes live on Majora actual. CS to try updated endpoint out on Thursday.

**Comment by cs46** [comment]
@SamStudio8 @tomrconnor today i have managed to write the script that will allow the sender_sample_id to be updated with the lab bar code.   There are 2 situations that need to be handled:

1. the sample does not have a biosample record (validity end point - exists = False)
2. the sample already has a blank biosample record (validity enpoint - exists = True, has_metadata = False, sender_sample_id = False)

For situation 1 i am able to create a blank biosample record and populate the sender sample ID without issue.  For situation 2 I am required to stomp the record and i am asked to enter the is_surveillance flag which i will not know.  I would suggest that the best course of action would be to allow a NULL into that field to avoid any confusion.  I would be wary of setting a default value as this could skew reporting.

My code is essentially ready to go and i would just need to add it to our cron jobs to run after every metadata upload attempt (every 1 hour during the day and once late at night to capture all of the days changes).

We also need to decide how far back we go.  I have used an arbitrary 01/01/2021 when sequencing was ramped up and available data quality increased.  I can go back further but i am not sure if this is necessary anyhow.  If we haven't linked it by now will we ever?

**Comment by SamStudio8** [comment]
You're saying in the second case you also know the adm1 and have a
collection_date (or a received_date, I guess)? I think we did previously
discuss lifting the requirement that is_surveillance is provided and I
don't see a good reason to keep enforcing it. I'd certainly prefer it to be
NULL rather than a guess.

I was wondering why there are so many of the (2) cases, but after
discussing this with @cs46 I understand that the add.library calls at WSI
are still using the force_biosamples option. In the first instance that
will need to be discussed internally before we can take action.
Ideally @cs46 would push the blank biosamples (with linkage) before the
add.library calls are made. Although I appreciate in practice it might
simply not be possible to retrieve the linkage before the sequence metadata
is uploaded like this. I'd certainly like to move away from WSI force
pushing all sequenced biosamples if at all possible, as that limits our
options currently.

**Comment by SamStudio8** [comment]
I have set up a `DA1` dataview on Majora actual with the configuration:

```
DA1
National orphaned identifiers
National orphaned identifiers
BiosampleArtifact
MF      BiosampleArtifact       central_sample_id
MF      BiosampleArtifact       sender_sample_id
MF      BiosourceSamplingProcess        adm1
MF      BiosourceSamplingProcess        collection_date
MF      BiosourceSamplingProcess        received_date
FF      created__biosourcesamplingprocess__collection_location_adm1     isnull  bool    1
```

This view will return the `central_sample_id` - `sender_sample_id` linkage (as well as any available dates) for samples where the adm1 is null (ie. blank samples). The DA1 view has been tested by PHW and is working as expected. The linkage that @cs46 pushed as a production test is visible to the DA :tada:

**Comment by SamStudio8** [comment]
This is obviously still work in progress as @cs46 and WSI need to work out how the biosample pushing process is going to work internally going forward (and we'll try and help with this where practical), but DAs can and should now get in touch to register access to the DA1 dataview to start rescuing orphan samples. The only work remaining will only add more linkage rather than more functionality.

**Comment by SamStudio8** [comment]
Granted force_add to MB to test the reciprocal linkage pushing from PHW to DA1. Will grant access to FS to read DA1 also. Having spoken to @cs46 there are two outstanding issues here:

1) WSI are routinely force pushing all biosamples when sending library metadata, this is legacy behaviour from when the delay between biosample metadata and sequencing at WSI was problematic. Nowadays WSI have their own metadata feed to assist with picking so potentially this behaviour could just be switched off (and CS/DJ will have to organise a protocol to ensure the biosamples exist before the sequencing). This is a problem as @cs46 can't force-push a sample with the linkage if the sample has been force-pushed already (it's a one time deal).
2) Secondly, for English samples where @cs46 has some metadata, it is not possible to push samples that do not have `is_surveillance` set as the field is mandatory. CS suggests that we lift the requirement for that field in the short term (and I am inclined to agree), medium term however I think CS should force-push the linkage for English samples and let PHE handle the linkage and pushing of the metadata record instead.

CS will chat to DJ internally to see what the options are re (1), then we can see what technical solutions could be offered in the case where it is not possible to arrange a protocol for CS to always upload biosamples first. For (2), I think we should remove the requirement for this field but will need sign off.

**Comment by cs46** [comment]
@SamStudio8 thanks for the write up.

Point 1 is the biggest hurdle for DA samples which i will need to talk with David Jackson about as he initiates the sequence upload. This is potentially a non-insignificant change and logistically could be problematic to put something in place quickly.  I was unaware that when a blank biosample is created by the api.artifact.library.add API call it would prevent me uploading the sender_sample_id.  This is the current process for every sequence and it is usually the first action conducted in Majora.  In this scenario I thought i would be be able to stomp with only the sender_sample_id populated.  It appears that once a blank record has been created then i cannot create another so i am unable to proceed.

For point 2, this is not an issue as i was confusing the blank and add end points.  I was receiving a "missing is_surveillance" error when i attempted to add a biosample record on top of an existing blank biosample which will obviously not work.  I initially tried creating a blank on top of the blank which results in a success but I didnt realise this action was not viable.

I will need to discuss with Frank how he wishes me to proceed with the population of sender_sample_id for english samples as this has not been in scope for my first iteration.

I will talk with David to see what is possible from our end.

I know its "hacky" and i appreciate that this may not go any further but if i was able to create a blank record with only a sender_sample_id on top of a blank record then my current solution would work for DA samples.

**Comment by cs46** [comment]
Changes made to the Sanger metadata upload process to also populate sender_sample_id if biosample is blank and we have data for the english sample.

David has confirmed that his process has not changed recently.

**Comment by cs46** [comment]
I have deployed my code to run every time we attempt to upload metadata.  This will not capture all samples due to the above mentioned problems but it will capture some so i thought it best to turn it on anyway.  We will need to discuss next steps to ensure this is working for all samples.

**Comment by SamStudio8** [comment]
CS reports via Operational Health Informatics that it will be too difficult to ensure the ordering of force pushes at WSI, so we'll proceed with looking at making the blank endpoint stomp over the `sender_sample_id` (rather than refusing to update it if the sample already exists). We briefly discussed this off-thread and it's a reasonable compromise. WSI should revisit whether they need to force push biosamples at the library level in the near future however, as this was not introduced to be normal day-to-day behaviour!

**Comment by cs46** [comment]
@SamStudio8 the biggest hurdle is that David's team does not have view of the LHL barcode and COPG-UK ID link and to get around this it would require significant re-engineering work out end.  The gold standard would be that the library add/upd is handled by my team and this wouldnt be an issue.  But this would be a few weeks work and then it would be competing with other scheduled work making the lead time into months.  This can be where we want to get to eventually so that workaround are no longer needed.

In the mean time i agree that the only way i see around this if we can stomp a blank biosample record with a blank biosample record and add in the sender sample ID at the second attempt.

**Comment by SamStudio8** [comment]
@cs46 Understood! I was more pushing for David's team to not force push biosamples at all (ie. the biosamples [with whatever metadata you want to provide] should already exist from your pipeline *before* the libraries get uploaded downstream). I was hoping it might just turn out that David could have turned this legacy behaviour off - you never know... As it happens what we have settled on was what I was already expecting to need to do, so I left the door open in my implementation to allow this to work ;)

**Comment by SamStudio8** [comment]
https://github.com/SamStudio8/majora/commit/e07201fe63d860bb971c0dda617d06ef22d9a41a allows the `sender_sample_id` to be updated by the `addempty` endpoint for the following cases:

* A completely new blank biosample
* A previously added blank biosample that has not yet had metadata uploaded to it through the full `add` endpoint
* A previously added blank biosample with sender_sample_id set by `addempty` that has not yet had metadata uploaded through `add`

Any successive call to `addempty` will update `sender_sample_id` if it is provided. Calling `addempty` on a sample without the `sender_sample_id` in the payload **will not** blank the field (but sending `null` will). @cs46 this should decouple your process from David's team.

This is deployed on Majora Magenta for testing but I don't see why we couldn't have this working on Majora Actual before the end of the week.

**Comment by cs46** [comment]
Successfully tested in Magenta, @SamStudio8 then deployed to prod.  Tested with one sample and it worked as expected. Modified the query back to the beginning of the month and worked as expected.  Changed to the beginning of 2021 and has worked as expected.

Blank biosamples with the LHL barcode in the sender_sample_id field will now be created if biosample record does not exist but sequence data will be uploaded.  When sequence data is already uploaded a blank biosample record will be stomped with the LHL barcode in sender_sample_id.

This now runs after every metadata upload process and should conclude this piece of work.

@tomrconnor Might i suggest that we monitor for the next week or so to ensure this has the desired effects?

**Comment by SamStudio8** [comment]
I think this is running smoothly now. @cs46 Any updates from your end? @tomrconnor This should probably be pushed to the 4 nations WG to organise access.

**Comment by cs46** [comment]
@SamStudio8 my code has been in place for several weeks now. We should be uploading the sender sample ID for all Sanger samples published to Majora. That being said, no one has formally signed it off as working but I haven’t had any complaints either. We really need to have a 3rd party check against the csv lists that there is parity.

**Comment by SamStudio8** [comment]
I think I'll close this as I'm satisfied our end of this work is done. We can re-open if something crops up.
